lib/elixir/lib/kernel.ex

```diff
@@ -780,8 +781,9 @@ defmodule Kernel do
 
       iex> tl([1, 2, 3, :go])
       [2, 3, :go]
+
   """
-  @spec tl(maybe_improper_list) :: maybe_improper_list
+  @spec tl(nonempty_maybe_improper_list) :: maybe_improper_list | term
```

Comment:
rather like `@spec tl(non_empty_maybe_improper_list(elem, tail) :: maybe_improper_list(elem, tail) | tail when elem: var, tail: var`. Using the variable means any inferred type information about the list itself remains. For example `tl([:a | :b, ...]) :: [:a | :b]`.